DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,237 @@
+
+  bool passed = true;
+
+  try {
+    queue q(device_selector, dpc_common::exception_handler);
+
+    constexpr int kVal1 = 1000, kVal2 = 2;
+
+    // Kernel `BasicOpsInt` contains native `int` type addition, multiplication,
+    // and division operations, while kernel `BasicOpsAcInt` contains `ac_int`
+    // type addition, multiplication, and division operations. By comparing
+    // these two kernels, you will find reduced width `ac_int` generates more
+    // efficient hardware than native `int`.
+    {
+      MyInt14 input_a = kVal1, input_b = kVal2;
+      MyInt15 output_c;
+      MyInt28 output_d;
+      MyInt15 output_e;
+      TestBasicOpsAcInt(q, input_a, input_b, output_c, output_d, output_e);
+
+      int golden_c, golden_d, golden_e;
+      TestBasicOpsInt(q, input_a, input_b, golden_c, golden_d, golden_e);
+
+      if (output_c != golden_c || output_d != golden_d ||
+          output_e != golden_e) {
+        std::cout << "Result mismatch!\n"
+                  << "Kernel BasicOpsInt:   addition = " << golden_c
+                  << ", multiplication = " << golden_d
+                  << ", division = " << golden_e << "\n"
+                  << "Kernel BasicOpsAcInt: addition = " << output_c
+                  << ", multiplication = " << output_d
+                  << ", division = " << output_e << "\n\n";
+        passed = false;
+      }
+    }
+
+    // Kernel `ShiftOps` contains an `ac_int` left shifter and the data type of
+    // the shift amount is a large width signed `ac_int`. On contrast, kernel
+    // `EfficientShiftOps` also contains an `ac_int` left shifter but the data
+    // type of the shift amount is a reduced width unsigned `ac_int`. By
+    // comparing these two kernels, you will find shift operations of `ac_int`
+    // can generate more efficient hardware if the amount to shift by is stored
+    // in a minimally sized unsigned `ac_int`.
+    {
+      MyInt14 input_a = kVal1, input_b = kVal2;
+      MyUInt2 input_efficient_b = kVal2;
+      MyInt14 output_c, output_efficient_c;
+      TestShiftOps(q, input_a, input_b, output_c);
+      TestEfficientShiftOps(q, input_a, input_efficient_b, output_efficient_c);
+
+      if (output_c != output_efficient_c) {
+        std::cout << "Result mismatch!\n"
+                  << "Kernel ShiftOps: result = " << output_c << "\n"
+                  << "Kernel EfficientShiftOps: result = " << output_efficient_c
+                  << "\n\n";
+        passed = false;
+      }
+    }
+
+    // Kernel `BitAccess` demonstrates bit access with bit select operator `[]`
+    // and bit slice write operation `set_slc`. Note: An `ac_int` must be
+    // initialized before being access by bit select operator `[]` and bit slice
+    // operations `slc` and `set_slc`, otherwise it is undefined behavior and
+    // will give you unexpected results.
+    {
+      MyInt14 input = kVal1;
+      MyInt14 output = TestBitAccess(q, input);
+
+      constexpr int golden = 0b001111101111;
```

Comment:
This is a case where you may want to use kGolden, as you are giving a name to a specific magic number.

Reply:
Will update.